File: backend/biz/handler/binding.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter

from application.expert_direct_relation import ExpertDirectRelationApplication

router = APIRouter(prefix="/binding")
application = ExpertDirectRelationApplication()


def _build_relation_response(
    relation_type: str,
    data_source: str = "all",
    expert_a_id: Optional[str] = None,
    expert_b_id: Optional[str] = None,
    institution: Optional[str] = None,
    start_time: Optional[str] = None,
    end_time: Optional[str] = None,
) -> dict[str, object]:
    return application.get_relation_response(
        data_source=data_source,
        expert_a_id=expert_a_id,
        expert_b_id=expert_b_id,
        institution=institution,
        relation_type=relation_type,
        start_time=start_time,
        end_time=end_time,
    )


@router.get("/expert-direct-relation")
async def expert_direct_relation(
    dataSource: str = "all",
    expertAId: Optional[str] = None,
    expertBId: Optional[str] = None,
    institution: Optional[str] = None,
    relationType: str = "direct",
    startTime: Optional[str] = None,
    endTime: Optional[str] = None,
) -> dict[str, object]:
    return _build_relation_response(
        relation_type=relationType,
        data_source=dataSource,
        expert_a_id=expertAId,
        expert_b_id=expertBId,
        institution=institution,
        start_time=startTime,
        end_time=endTime,
    )


@router.get("/expert-direct-two-hop")
async def expert_direct_two_hop(
    dataSource: str = "all",
    expertAId: Optional[str] = None,
    expertBId: Optional[str] = None,
    institution: Optional[str] = None,
    startTime: Optional[str] = None,
    endTime: Optional[str] = None,
) -> dict[str, object]:
    return _build_relation_response(
        relation_type="two_hop",
        data_source=dataSource,
        expert_a_id=expertAId,
        expert_b_id=expertBId,
        institution=institution,
        start_time=startTime,
        end_time=endTime,
    )


@router.get("/expert-direct-three-hop")
async def expert_direct_three_hop(
    dataSource: str = "all",
    expertAId: Optional[str] = None,
    expertBId: Optional[str] = None,
    institution: Optional[str] = None,
    startTime: Optional[str] = None,
    endTime: Optional[str] = None,
) -> dict[str, object]:
    return _build_relation_response(
        relation_type="three_hop",
        data_source=dataSource,
        expert_a_id=expertAId,
        expert_b_id=expertBId,
        institution=institution,
        start_time=startTime,
        end_time=endTime,
    )

File: frontend/src/api/binding.ts
export type BindingRelationType = 'direct' | 'two_hop' | 'three_hop'

export interface BindingQueryParams {
  dataSource?: string
  expertAId?: string
  expertBId?: string
  institution?: string
  relationType?: BindingRelationType
  startTime?: string
  endTime?: string
}

export interface BindingScenario {
  key?: string
  label?: string
  last_test_time?: string
  expert_a?: { id?: string; name?: string; title?: string }
  expert_b?: { id?: string; name?: string; title?: string }
  relation_type?: string
  institution?: string
  directions?: string[]
  api_example?: Record<string, unknown>
}

export async function fetchBindingScenarios(params: BindingQueryParams): Promise<{ scenarios: BindingScenario[] }> {
  const endpoint =
    params.relationType === 'two_hop'
      ? '/api/v1/binding/expert-direct-two-hop'
      : params.relationType === 'three_hop'
        ? '/api/v1/binding/expert-direct-three-hop'
        : '/api/v1/binding/expert-direct-relation'

  const search = new URLSearchParams()
  if (params.dataSource) search.set('dataSource', params.dataSource)
  if (params.expertAId) search.set('expertAId', params.expertAId)
  if (params.expertBId) search.set('expertBId', params.expertBId)
  if (params.institution) search.set('institution', params.institution)
  if (params.relationType && params.relationType !== 'direct') search.set('relationType', params.relationType)
  if (params.startTime) search.set('startTime', params.startTime)
  if (params.endTime) search.set('endTime', params.endTime)

  const response = await fetch(`${endpoint}?${search.toString()}`)
  if (!response.ok) {
    const body = await response.json().catch(() => null)
    throw new Error(body?.detail || `接口请求失败：${response.status}`)
  }
  return response.json()
}
